In [2]:
import random
from funcoes import (
    criar_gene,
    cria_populacao,
    calcular_aptidao,
    funcao_objetivo_pop,
    selecao_torneio,
    cruzamento_ponto_simples_perm,
    mutacao_troca
)

# Parâmetros do algoritmo
TAMANHO_POPULACAO = 100
NUM_GERACOES = 500
CHANCE_DE_CRUZAMENTO = 0.7
CHANCE_DE_MUTACAO = 0.01

PACIENTES = [
    {"id": 1, "idade": 70, "genero": "F", "comorbidades": 3},
    {"id": 2, "idade": 65, "genero": "M", "comorbidades": 2},
    {"id": 3, "idade": 45, "genero": "F", "comorbidades": 3},
    {"id": 4, "idade": 46, "genero": "M", "comorbidades": 0},
    {"id": 5, "idade": 50, "genero": "M", "comorbidades": 1},
    {"id": 6, "idade": 80, "genero": "F", "comorbidades": 4},
    {"id": 7, "idade": 55, "genero": "M", "comorbidades": 0},
    {"id": 8, "idade": 60, "genero": "F", "comorbidades": 2},
    {"id": 9, "idade": 65, "genero": "M", "comorbidades": 3},
    {"id": 10, "idade": 70, "genero": "F", "comorbidades": 1},
    {"id": 11, "idade": 75, "genero": "M", "comorbidades": 2},
    # Adicione mais pacientes conforme necessário
]

def algoritmo_genetico(pacientes):
    """
    Executa o algoritmo genético para otimizar a ordem de atendimento dos pacientes.

    Args:
        pacientes (list): Lista de dicionários representando os pacientes.

    Returns:
        tuple: Melhor indivíduo observado e seu valor de aptidão.
    """
    hall_da_fama = []
    populacao = cria_populacao(TAMANHO_POPULACAO, pacientes)

    for _ in range(NUM_GERACOES):
        fitness = funcao_objetivo_pop(populacao)
        selecionados = selecao_torneio(populacao, fitness)
        
        proxima_geracao = []
        for pai, mae in zip(selecionados[::2], selecionados[1::2]):
            filho1, filho2 = cruzamento_ponto_simples_perm(pai, mae, CHANCE_DE_CRUZAMENTO)
            proxima_geracao.append(filho1)
            proxima_geracao.append(filho2)
        
        for individuo in proxima_geracao:
            mutacao_troca(individuo, CHANCE_DE_MUTACAO)
        
        fitness = funcao_objetivo_pop(proxima_geracao)
        maior_fitness = max(fitness)
        indice = fitness.index(maior_fitness)
        hall_da_fama.append(proxima_geracao[indice])
        
        populacao = proxima_geracao
    
    fitness = funcao_objetivo_pop(hall_da_fama)
    maior_fitness = max(fitness)
    indice = fitness.index(maior_fitness)
    melhor_individuo_observado = hall_da_fama[indice]
    
    return melhor_individuo_observado, maior_fitness

melhor_individuo, maior_fitness = algoritmo_genetico(PACIENTES)
print("Melhor indivíduo observado:")
for paciente in melhor_individuo:
    print(paciente)
print("Maior fitness:", maior_fitness)


Melhor indivíduo observado:
{'id': 6, 'idade': 80, 'genero': 'F', 'comorbidades': 4}
{'id': 11, 'idade': 75, 'genero': 'M', 'comorbidades': 2}
{'id': 1, 'idade': 70, 'genero': 'F', 'comorbidades': 3}
{'id': 9, 'idade': 65, 'genero': 'M', 'comorbidades': 3}
{'id': 10, 'idade': 70, 'genero': 'F', 'comorbidades': 1}
{'id': 2, 'idade': 65, 'genero': 'M', 'comorbidades': 2}
{'id': 8, 'idade': 60, 'genero': 'F', 'comorbidades': 2}
{'id': 3, 'idade': 45, 'genero': 'F', 'comorbidades': 3}
{'id': 7, 'idade': 55, 'genero': 'M', 'comorbidades': 0}
{'id': 5, 'idade': 50, 'genero': 'M', 'comorbidades': 1}
{'id': 4, 'idade': 46, 'genero': 'M', 'comorbidades': 0}
Maior fitness: 194.5231637806638


## Teste com Diferentes Filas
Vamos testar o algoritmo com diferentes filas, incluindo filas improváveis e filas bastante prováveis, para observar as mudanças:

In [4]:
# Definindo diferentes filas para teste
fila_improvavel = [
    {"id": 1, "idade": 70, "genero": "F", "comorbidades": 0},
    {"id": 2, "idade": 20, "genero": "M", "comorbidades": 0},
    {"id": 3, "idade": 45, "genero": "F", "comorbidades": 0},
    {"id": 4, "idade": 30, "genero": "M", "comorbidades": 0},
    {"id": 5, "idade": 50, "genero": "M", "comorbidades": 0},
    {"id": 6, "idade": 80, "genero": "F", "comorbidades": 0},
    {"id": 7, "idade": 35, "genero": "M", "comorbidades": 0},
    {"id": 8, "idade": 60, "genero": "F", "comorbidades": 0},
    {"id": 9, "idade": 65, "genero": "M", "comorbidades": 0},
    {"id": 10, "idade": 70, "genero": "F", "comorbidades": 0},
    {"id": 11, "idade": 75, "genero": "M", "comorbidades": 0},
]

fila_provavel = [
    {"id": 1, "idade": 80, "genero": "F", "comorbidades": 4},
    {"id": 2, "idade": 75, "genero": "M", "comorbidades": 3},
    {"id": 3, "idade": 70, "genero": "F", "comorbidades": 3},
    {"id": 4, "idade": 65, "genero": "M", "comorbidades": 2},
    {"id": 5, "idade": 60, "genero": "F", "comorbidades": 2},
    {"id": 6, "idade": 55, "genero": "M", "comorbidades": 1},
    {"id": 7, "idade": 50, "genero": "M", "comorbidades": 1},
    {"id": 8, "idade": 45, "genero": "F", "comorbidades": 0},
    {"id": 9, "idade": 40, "genero": "M", "comorbidades": 0},
    {"id": 10, "idade": 35, "genero": "F", "comorbidades": 0},
    {"id": 11, "idade": 30, "genero": "M", "comorbidades": 0},
]

print("Teste com fila improvável:")
melhor_individuo, maior_fitness = algoritmo_genetico(fila_improvavel)
print("Melhor indivíduo observado:")
for paciente in melhor_individuo:
    print(paciente)
print("Maior fitness:", maior_fitness)

print("\nTeste com fila provável:")
melhor_individuo, maior_fitness = algoritmo_genetico(fila_provavel)
print("Melhor indivíduo observado:")
for paciente in melhor_individuo:
    print(paciente)
print("Maior fitness:", maior_fitness)


Teste com fila improvável:
Melhor indivíduo observado:
{'id': 6, 'idade': 80, 'genero': 'F', 'comorbidades': 0}
{'id': 11, 'idade': 75, 'genero': 'M', 'comorbidades': 0}
{'id': 10, 'idade': 70, 'genero': 'F', 'comorbidades': 0}
{'id': 1, 'idade': 70, 'genero': 'F', 'comorbidades': 0}
{'id': 9, 'idade': 65, 'genero': 'M', 'comorbidades': 0}
{'id': 8, 'idade': 60, 'genero': 'F', 'comorbidades': 0}
{'id': 5, 'idade': 50, 'genero': 'M', 'comorbidades': 0}
{'id': 3, 'idade': 45, 'genero': 'F', 'comorbidades': 0}
{'id': 7, 'idade': 35, 'genero': 'M', 'comorbidades': 0}
{'id': 4, 'idade': 30, 'genero': 'M', 'comorbidades': 0}
{'id': 2, 'idade': 20, 'genero': 'M', 'comorbidades': 0}
Maior fitness: 163.52757215007216

Teste com fila provável:
Melhor indivíduo observado:
{'id': 1, 'idade': 80, 'genero': 'F', 'comorbidades': 4}
{'id': 2, 'idade': 75, 'genero': 'M', 'comorbidades': 3}
{'id': 3, 'idade': 70, 'genero': 'F', 'comorbidades': 3}
{'id': 4, 'idade': 65, 'genero': 'M', 'comorbidades': 2}


## A partir dos resultados dos testes com diferentes filas, podemos fazer as seguintes observações:

- **Ordem dos Pacientes:** O algoritmo genético conseguiu encontrar uma ordem de atendimento que maximiza a aptidão total para ambas as filas. Em ambos os casos, os pacientes com maior idade e mais comorbidades foram priorizados no início da fila, o que está alinhado com a função de aptidão definida.

## Diferença entre Filas Improváveis e Prováveis:
- **Fila Improvável:** Esta fila contém pacientes com pouca ou nenhuma comorbidade, independentemente da idade. O melhor indivíduo observado priorizou pacientes mais velhos, já que as comorbidades são iguais para todos (zero). Isso resulta em uma aptidão relativamente alta, mas não ideal, dada a falta de variação significativa nos dados de comorbidade.

- **Fila Provável:** Esta fila contém pacientes com uma combinação de idade avançada e comorbidades. O melhor indivíduo observado neste caso reflete uma ordem de atendimento mais realista, onde pacientes mais velhos com mais comorbidades são priorizados. A aptidão resultante é maior, indicando uma melhor solução otimizada.

## Maior Aptidão:
- **Fila Improvável:** A maior aptidão foi 163.52757215007216. Isso sugere que, apesar da falta de variação nas comorbidades, o algoritmo conseguiu encontrar uma ordem razoável priorizando pacientes mais velhos.

- **Fila Provável:** A maior aptidão foi 185.73786075036077. Esta aptidão mais alta reflete uma otimização mais eficaz, já que há mais variabilidade nos dados (tanto em idade quanto em comorbidades), permitindo que o algoritmo encontre uma solução mais robusta.

## Eficiência do Algoritmo: 
Os resultados indicam que o algoritmo genético é eficaz em encontrar ordens de atendimento otimizadas, mesmo quando as filas têm características diferentes. A capacidade de ajustar a ordem com base em variáveis múltiplas (idade, gênero, comorbidades) mostra a flexibilidade e robustez do algoritmo.

## Impacto da Variabilidade nos Dados:
A maior variabilidade nos dados (como na fila provável) permite que o algoritmo encontre soluções com aptidão significativamente maior. Isso destaca a importância de considerar múltiplos fatores relevantes (como idade e comorbidades) para obter melhores resultados.

## Comparação com a Função de Aptidão: 
A função de aptidão usada é essencial para guiar o algoritmo genético a soluções desejáveis. Aqui, a função de aptidão baseada em idade e comorbidades parece ser adequada para priorizar pacientes de acordo com a urgência clínica.

Essas observações reforçam a utilidade do algoritmo genético para otimização de filas de atendimento em sistemas de saúde, especialmente quando os dados são diversos e refletem a complexidade do ambiente real.

In [9]:
import csv
import random
import os

# Função para gerar pacientes aleatórios
def gerar_pacientes_aleatorios(n):
    pacientes = []
    for i in range(1, n + 1):
        paciente = {
            "id": i,
            "idade": random.randint(20, 90),
            "genero": random.choice(["M", "F"]),
            "comorbidades": random.randint(0, 5)
        }
        pacientes.append(paciente)
    return pacientes

# Gerar 150 pacientes aleatórios
pacientes_aleatorios = gerar_pacientes_aleatorios(150)

# Definir o caminho e nome do arquivo CSV
csv_file_name = 'pacientes.csv'
csv_file_path = os.path.join(os.getcwd(), csv_file_name)

# Escrever os pacientes em um arquivo CSV
with open(csv_file_path, mode='w', newline='') as file:
    writer = csv.DictWriter(file, fieldnames=["id", "idade", "genero", "comorbidades"])
    writer.writeheader()
    writer.writerows(pacientes_aleatorios)

csv_file_path


'C:\\Users\\vinicius23011\\Downloads\\SUSOptiLine\\pacientes.csv'

In [11]:
import pandas as pd

def carregar_pacientes_de_csv(caminho_do_arquivo):
    """
    Carrega dados dos pacientes de um arquivo CSV.
    
    Args:
        caminho_do_arquivo (str): Caminho para o arquivo CSV.
    
    Returns:
        list: Lista de dicionários representando os pacientes.
    """
    df = pd.read_csv(caminho_do_arquivo)
    pacientes = df.to_dict(orient='records')
    return pacientes

# Exemplo de uso
caminho_do_arquivo = "pacientes.csv"
pacientes = carregar_pacientes_de_csv(caminho_do_arquivo)
melhor_individuo, maior_fitness = algoritmo_genetico(pacientes)

In [13]:
melhor_individuo, maior_fitness

([{'id': 39, 'idade': 90, 'genero': 'F', 'comorbidades': 5},
  {'id': 128, 'idade': 90, 'genero': 'M', 'comorbidades': 4},
  {'id': 14, 'idade': 84, 'genero': 'F', 'comorbidades': 5},
  {'id': 7, 'idade': 87, 'genero': 'F', 'comorbidades': 4},
  {'id': 89, 'idade': 88, 'genero': 'F', 'comorbidades': 4},
  {'id': 97, 'idade': 81, 'genero': 'M', 'comorbidades': 5},
  {'id': 34, 'idade': 86, 'genero': 'M', 'comorbidades': 4},
  {'id': 85, 'idade': 83, 'genero': 'F', 'comorbidades': 4},
  {'id': 54, 'idade': 84, 'genero': 'F', 'comorbidades': 3},
  {'id': 69, 'idade': 83, 'genero': 'F', 'comorbidades': 4},
  {'id': 55, 'idade': 85, 'genero': 'M', 'comorbidades': 4},
  {'id': 110, 'idade': 79, 'genero': 'F', 'comorbidades': 5},
  {'id': 46, 'idade': 78, 'genero': 'F', 'comorbidades': 5},
  {'id': 15, 'idade': 90, 'genero': 'M', 'comorbidades': 0},
  {'id': 43, 'idade': 86, 'genero': 'F', 'comorbidades': 2},
  {'id': 109, 'idade': 70, 'genero': 'F', 'comorbidades': 5},
  {'id': 112, 'idade':